src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,154 @@
+        if (index == 0) {
+            first.prev = newNode;
+            newNode.next = first;
+            first = newNode;
+            size++;
+            return;
+        }
+        Node<T> currentNode = findNodeByIndex(index);
+        if (currentNode.next == null) {
+            T lastItem = currentNode.item;
+            currentNode.item = value;
+            add(lastItem);
+        } else {
+            currentNode.prev.next = newNode;
+            newNode.next = currentNode;
+            newNode.prev = currentNode.prev;
+            currentNode.prev = newNode;
+            size++;
+        }
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        return currentNode.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        T oldValue = currentNode.item;
+        currentNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeToRemove = findNodeByIndex(index);
+        unlink(nodeToRemove);
+        return nodeToRemove.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> objectToRemove = first;
+        for (int i = 0; i < size; i++) {
+            if (object == objectToRemove.item
+                    || (object != null && object.equals(objectToRemove.item))) {
+                unlink(objectToRemove);
+                return true;
+            }
+            objectToRemove = objectToRemove.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unlink(Node<T> nodeToRemove) {
+        if (nodeToRemove == first && first.next == null) {
+            first = null;
+        } else if (nodeToRemove == first) {
+            first.next.prev = null;
+            first = first.next;
+        } else if (nodeToRemove == last) {
+            nodeToRemove.prev.next = null;
+            nodeToRemove.prev = last;
+        } else {
+            nodeToRemove.next.prev = nodeToRemove.prev;
+            nodeToRemove.prev.next = nodeToRemove.next;
+        }
+        size--;
+    }
+
+    private void checkIndexAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayIndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS_MESSAGE);
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayIndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS_MESSAGE);
```

Comment:
include index and size in the message, use String.format if you want to use constant